src/collider/GeomTree.cpp

```diff
@@ -189,90 +203,34 @@ GeomTree::GeomTree(Serializer::Reader &rd)
 		m_triFlags[iTri] = rd.Int32();
 	}
 
-	// activeTris = tris we are still trying to put into leaves
-	std::vector<int> activeTris;
-	activeTris.reserve(m_numTris);
-	for (int i = 0; i < m_numTris; i++) {
-		activeTris.push_back(i * 3);
-	}
+	{
+		PROFILE_SCOPED_DESC("GeomTree::CreateTriTree");
+
+		// TODO: triangle AABBs should be written to the SGM file similarly to edge AABBs
+		m_triAABBs.reset(new AABBd[m_numTris]);
+		for (size_t i = 0; i < m_numTris; i++) {
+			const vector3d v0 = vector3d(m_vertices[m_indices[i * 3 + 0]]);
+			const vector3d v1 = vector3d(m_vertices[m_indices[i * 3 + 1]]);
+			const vector3d v2 = vector3d(m_vertices[m_indices[i * 3 + 2]]);
+
+			AABBd aabb = { v0, v0 };
+			aabb.Update(v1);
+			aabb.Update(v2);
+			m_triAABBs.get()[i] = aabb;
+		}
 
-	// regenerate the aabb data
-	Aabb *aabbs = new Aabb[activeTris.size()];
-	for (Uint32 i = 0; i < activeTris.size(); i++) {
-		const vector3d v1 = vector3d(m_vertices[m_indices[activeTris[i] + 0]]);
-		const vector3d v2 = vector3d(m_vertices[m_indices[activeTris[i] + 1]]);
-		const vector3d v3 = vector3d(m_vertices[m_indices[activeTris[i] + 2]]);
-		aabbs[i].min = aabbs[i].max = v1;
-		aabbs[i].Update(v2);
-		aabbs[i].Update(v3);
-	}
-	m_triTree.reset(new BVHTree(activeTris.size(), &activeTris[0], aabbs));
-	delete[] aabbs;
-
-	//
-	int *edgeIdxs = new int[m_numEdges];
-	memset(edgeIdxs, 0, sizeof(int) * m_numEdges);
-	for (int i = 0; i < m_numEdges; i++) {
-		edgeIdxs[i] = i;
+		m_triTree.reset(new BinnedAreaBVHTree());
+		m_triTree->Build(bounds, m_triAABBs.get(), m_numTris);
 	}
-
-	m_edgeTree.reset(new BVHTree(m_numEdges, edgeIdxs, &m_aabbs[0]));
-	delete[] edgeIdxs;
-}
-
-static bool SlabsRayAabbTest(const BVHNode *n, const vector3f &start, const vector3f &invDir, isect_t *isect)
-{
-	// PROFILE_SCOPED()
-	float
-		l1 = (n->aabb.min.x - start.x) * invDir.x,
-		l2 = (n->aabb.max.x - start.x) * invDir.x,
-		lmin = std::min(l1, l2),
-		lmax = std::max(l1, l2);
-
-	l1 = (n->aabb.min.y - start.y) * invDir.y;
-	l2 = (n->aabb.max.y - start.y) * invDir.y;
-	lmin = std::max(std::min(l1, l2), lmin);
-	lmax = std::min(std::max(l1, l2), lmax);
-
-	l1 = (n->aabb.min.z - start.z) * invDir.z;
-	l2 = (n->aabb.max.z - start.z) * invDir.z;
-	lmin = std::max(std::min(l1, l2), lmin);
-	lmax = std::min(std::max(l1, l2), lmax);
-
-	return ((lmax >= 0.f) & (lmax >= lmin) & (lmin < isect->dist));
 }
 
 void GeomTree::TraceRay(const vector3f &start, const vector3f &dir, isect_t *isect) const
 {
-	TraceRay(m_triTree->GetRoot(), start, dir, isect);
-}
+	std::vector<uint32_t> tri_isect;
+	m_triTree->TraceRay(vector3d(start), vector3d(1.f / dir), FLT_MAX, tri_isect);
```

Comment:
`dir` components can be 0.

Reply:
See https://tavianator.com/2011/ray_box.html - this is intentional and well defined on IEEE754-conforming implementations (i.e. every modern processor today). I've removed ReciprocalSafe() as a result, since all divide-by-zero occurrences are in fact intended.